### PR TITLE
Build with kernel v4.5.0+

### PR DESCRIPTION
**Pull request by clachan**
Hi,

I made some changes to nova to build it with kernel 4.5 and 4.6. Tested with v4.4, v4.5 and v4.6-rc4. Please review it.

Thanks,
Clay Chang

### CODE REVIEW

**Comment by Andiry** [comment]
Thank you for submitting this, I will review it.

File: dax.c [hunk 1]
```diff
@@ -16,6 +16,10 @@
 #include <linux/buffer_head.h>
 #include <asm/cpufeature.h>
 #include <asm/pgtable.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 5, 0)
+#include <linux/pfn_t.h>
+#endif
```
Comment (Andiry) [inline dax.c]: Move these includes to nova.h. We don't need them in each file.

File: super.c [hunk 2]
```diff
@@ -89,7 +93,11 @@ static int nova_get_block_info(struct super_block *sb,
 	struct nova_sb_info *sbi)
 {
 	void *virt_addr = NULL;
+#if LINUX_KERNEL_VERSION >= KERNEL_VERSION(4, 5, 0)
```
Comment (Andiry) [inline super.c]: LINUX_KERNEL_VERSION -> LINUX_VERSION_CODE

File: super.c [hunk 3]
```diff
@@ -99,11 +107,20 @@ static int nova_get_block_info(struct super_block *sb,
...
+	size = sb->s_bdev->bd_disk->fops->direct_access(sb->s_bdev,
+					0, &virt_addr, &__pfn_t);
+#else
 	size = sb->s_bdev->bd_disk->fops->direct_access(sb->s_bdev,
 					0, &virt_addr, &pfn);
+#endif
 
 	sbi->virt_addr = virt_addr;
+#if LINUX_KERNEL_VERSION >= KERNEL_VERSION(4, 5, 0)
+	sbi->phys_addr = __pfn_t.val << PAGE_SHIFT;
```
Comment (Andiry) [inline super.c]: Let's use pfn_t_to_pfn() here.

File: symlink.c [hunk 4]
```diff
@@ -80,10 +81,16 @@ static int nova_readlink(struct dentry *dentry, char __user *buffer, int buflen)
...
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 5, 0)
+static const char *nova_get_link(struct dentry *dentry, struct inode *inode, void **cookie)
+#else
 static const char *nova_follow_link(struct dentry *dentry, void **cookie)
+#endif
 {
 	struct nova_file_write_entry *entry;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 5, 0)
 	struct inode *inode = dentry->d_inode;
+#endif
```
Comment (Andiry) [inline symlink.c]: This looks wield. Can you do it this way:

nova_get_link(){
...
}

+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 5, 0)
+nova_follow_link() {
  ...
  return nova_get_link();
}
+#else

**Comment by clachan** [comment]
Hi Andiry. I've added some commits to address your comments, please review them.

Thanks,
Clay

**Comment by Andiry** [comment]
Merged, thanks.

**Pull request merged**